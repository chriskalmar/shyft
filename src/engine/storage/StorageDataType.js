import { passOrThrow, isFunction, isArray } from '../util';

import constants from '../constants';

export class StorageDataType {
  constructor(setup = {}) {
    const {
      name,
      description,
      nativeDataType,
      isSortable,
      serialize,
      enforceSerialize,
      parse,
      capabilities,
    } = setup;

    passOrThrow(name, () => 'Missing storage data type name');
    passOrThrow(
      description,
      () => `Missing description for storage data type '${name}'`,
    );
    passOrThrow(
      nativeDataType,
      () => `Missing native data type for storage data type '${name}'`,
    );

    passOrThrow(
      isFunction(serialize),
      () => `Storage data type '${name}' has an invalid serialize function`,
    );

    passOrThrow(
      !parse || isFunction(parse),
      () => `Storage data type '${name}' has an invalid parse function`,
    );

    if (capabilities) {
      passOrThrow(
        isArray(capabilities),
        () => `Storage data type '${name}' has an invalid list of capabilities`,
      );

      capabilities.map(capability => {
        passOrThrow(
          constants.storageDataTypeCapabilities[capability],
          () =>
            `Storage data type '${name}' has an unknown capability '${capability}'`,
        );
      });
    }

    this.name = name;
    this.description = description;
    this.nativeDataType = nativeDataType;
    this.isSortable = !!isSortable;
    this.serialize = serialize;
    this.enforceSerialize = !!enforceSerialize;
    this.parse = parse || (value => value);
    this.capabilities = capabilities || [];
  }

  toString() {
    return this.name;
  }
}

export const isStorageDataType = obj => {
  return obj instanceof StorageDataType;
};
