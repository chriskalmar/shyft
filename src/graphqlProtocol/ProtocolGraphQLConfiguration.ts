import {
  generateTypeName,
  generateTypeNamePlural,
  generateTypeNamePascalCase,
  generateTypeNamePluralPascalCase,
  generateTypeNameUpperCase,
} from './util';
import { MAX_PAGE_SIZE } from './protocolGraphqlConstants';
import { ProtocolConfiguration } from '../engine/protocol/ProtocolConfiguration';
import { Mutation, Subscription, Entity, Action, ViewEntity } from '..';
import { ShadowEntity } from '../engine/entity/ShadowEntity';

export class ProtocolGraphQLConfiguration extends ProtocolConfiguration {
  constructor() {
    super();

    this.enableFeatures(
      [
        'query',
        'action',
        'mutation',
        'mutationNested',
        'mutationById',
        'mutationByUniqueness',
        'subscription',
        'subscriptionNested',
        'subscriptionById',
      ],
      true,
    );
  }

  getMaxPageSize() {
    return Number(process.env.MAX_PAGE_SIZE) || MAX_PAGE_SIZE;
  }

  /* name generators */

  generateEntityTypeName(entity: Entity | ViewEntity | ShadowEntity): string {
    return generateTypeName(entity.name);
  }

  generateEntityTypeNamePlural(
    entity: Entity | ViewEntity | ShadowEntity,
  ): string {
    return generateTypeNamePlural(entity.name);
  }

  generateEntityTypeNamePascalCase(
    entity: Entity | ViewEntity | ShadowEntity,
  ): string {
    return generateTypeNamePascalCase(entity.name);
  }

  generateEntityTypeNamePluralPascalCase(
    entity: Entity | ViewEntity | ShadowEntity,
  ): string {
    return generateTypeNamePluralPascalCase(entity.name);
  }

  generateReverseConnectionFieldName(
    sourceEntity: Entity | ViewEntity | ShadowEntity,
    sourceAttributeName,
  ): string {
    const typeNamePlural = this.generateEntityTypeNamePlural(sourceEntity);
    return generateTypeName(`${typeNamePlural}-by-${sourceAttributeName}`);
  }

  generateReferenceFieldName(
    referenceEntity: Entity | ViewEntity | ShadowEntity,
    attribute,
  ): string {
    const fieldName = this.generateFieldName(attribute);
    return generateTypeName(`${referenceEntity.name}-by-${fieldName}`);
  }

  generateReferenceFieldListName(referenceEntity, attribute): string {
    const fieldName = this.generateFieldName(attribute);
    const referenceEntityNamePlural = this.generateEntityTypeNamePlural(
      referenceEntity,
    );
    return generateTypeName(`${referenceEntityNamePlural}-by-${fieldName}`);
  }

  generateFieldName(attribute): string {
    return generateTypeName(attribute.name);
  }

  generateI18nJsonFieldName(attribute): string {
    const fieldName = this.generateFieldName(attribute);
    return `${fieldName}_i18nJson`;
  }

  generateI18nFieldName(attribute): string {
    const fieldName = this.generateFieldName(attribute);
    return `${fieldName}_i18n`;
  }

  generateI18nFieldTypeName(entity: Entity | ViewEntity, attribute) {
    const typeName = this.generateEntityTypeName(entity);
    const fieldName = this.generateFieldName(attribute);
    return generateTypeNamePascalCase(`${typeName}-${fieldName}-i18n`);
  }

  generateListQueryTypeName(entity: Entity | ViewEntity) {
    const typeNamePlural = this.generateEntityTypeNamePlural(entity);
    return generateTypeName(`all-${typeNamePlural}`);
  }

  generateInstanceQueryTypeName(entity: Entity | ViewEntity): string {
    return this.generateEntityTypeName(entity);
  }

  generateInstanceByUniqueQueryTypeName(
    entity: Entity | ViewEntity,
    attribute,
  ): string {
    const typeName = this.generateEntityTypeName(entity);
    const fieldName = this.generateFieldName(attribute);
    return generateTypeName(`${typeName}-by-${fieldName}`);
  }

  generateUniquenessAttributesName(
    _entity: Entity | ViewEntity,
    attributes,
  ): string {
    return generateTypeName(attributes.join('-and-'));
  }

  generateUniquenessAttributesFieldName(
    _entity: Entity | ViewEntity,
    attribute,
    uniquenessAttributesName: string,
  ): string {
    const fieldName = this.generateFieldName(attribute);
    return generateTypeName(
      `${fieldName}-by-unique-${uniquenessAttributesName}`,
    );
  }

  generateInstanceUniquenessInputTypeName(
    entity: Entity | ViewEntity,
    uniquenessAttributesName,
  ): string {
    const typeName = this.generateEntityTypeName(entity);
    return generateTypeNamePascalCase(
      `${typeName}-instance-uniqueness-on-${uniquenessAttributesName}-input`,
    );
  }

  generateActionTypeName(action: Action): string {
    return generateTypeName(action.name);
  }

  generateDataInputTypeName(baseName: string): string {
    return generateTypeNamePascalCase(`${baseName}-data-input`);
  }

  generateNestedDataInputTypeName(
    baseName: string,
    nestedParamName: string,
    level: number,
  ): string {
    const levelStr = level > 1 ? `L${level}` : '';

    return generateTypeNamePascalCase(
      `${baseName}-${nestedParamName}-${levelStr}-data-input`,
    );
  }

  generateInputTypeName(baseName: string): string {
    return generateTypeNamePascalCase(`${baseName}-input`);
  }

  generateDataOutPutTypeName(baseName: string): string {
    return generateTypeNamePascalCase(`${baseName}-data-output`);
  }

  generateNestedDataOutPutTypeName(
    baseName: string,
    nestedParamName: string,
    level?: number,
  ): string {
    const levelStr = level > 1 ? `L${level}` : '';

    return generateTypeNamePascalCase(
      `${baseName}-${nestedParamName}-${levelStr}-data-output`,
    );
  }

  generateOutPutTypeName(baseName: string): string {
    return generateTypeNamePascalCase(`${baseName}-output`);
  }

  generateSortKeyName(attribute: any, ascending?: boolean): string {
    const direction = ascending ? 'ASC' : 'DESC';

    return `${generateTypeNameUpperCase(attribute.name)}_${direction}`;
  }

  generateSortInputTypeName(entity: Entity | ViewEntity): string {
    const typeName = this.generateEntityTypeName(entity);
    return generateTypeNamePascalCase(`${typeName}-order-by`);
  }

  generateFilterInputTypeName(entity: Entity | ViewEntity): string {
    const typeName = this.generateEntityTypeName(entity);
    return generateTypeNamePascalCase(`${typeName}-filter`);
  }

  generateFilterPreFilterInputTypeName(entity: Entity | ViewEntity): string {
    const typeName = this.generateEntityTypeName(entity);
    return generateTypeNamePascalCase(`${typeName}-pre-filter`);
  }

  generateFilterPreFilterParamsInputTypeName(
    entity: Entity | ViewEntity,
    preFilter,
  ): string {
    const typeName = this.generateEntityTypeName(entity);
    return generateTypeNamePascalCase(
      `${typeName}-pre-filter-${preFilter}-params`,
    );
  }

  generateConnectionEdgeTypeName(entity: Entity | ViewEntity): string {
    const typeName = this.generateEntityTypeName(entity);
    return generateTypeNamePascalCase(`${typeName}-edge`);
  }

  generateConnectionTypeName(entity: Entity | ViewEntity): string {
    const typeName = this.generateEntityTypeName(entity);
    return generateTypeNamePascalCase(`${typeName}-connection`);
  }

  generateOperationI18nAttributeInputTypeName(
    entity: Entity | any,
    operation: Mutation | Subscription | any,
    attribute,
  ): string {
    const typeName = this.generateEntityTypeName(entity);
    const fieldName = this.generateFieldName(attribute);
    return generateTypeNamePascalCase(
      `${operation.name}-${typeName}-${fieldName}-i18n-input`,
    );
  }

  generateOperationInstanceInputTypeName(
    entity: Entity,
    operation: Mutation | Subscription,
  ): string {
    const typeName = this.generateEntityTypeName(entity);
    return generateTypeNamePascalCase(
      `${operation.name}-${typeName}-instance-input`,
    );
  }

  generateOperationInputTypeName(
    entity: Entity,
    operation: Mutation | Subscription,
  ): string {
    const typeName = this.generateEntityTypeName(entity);
    return generateTypeNamePascalCase(`${operation.name}-${typeName}-input`);
  }

  generateOperationByPrimaryAttributeInputTypeName(
    entity: Entity,
    operation: Mutation | Subscription,
    attribute,
  ): string {
    const typeName = this.generateEntityTypeName(entity);
    const fieldName = this.generateFieldName(attribute);
    return generateTypeNamePascalCase(
      `${operation.name}-${typeName}-by-${fieldName}-input`,
    );
  }

  generateOperationInstanceNestedInputTypeName(
    entity: Entity,
    operation: Mutation | Subscription,
  ): string {
    const typeName = this.generateEntityTypeName(entity);
    return generateTypeNamePascalCase(
      `${operation.name}-${typeName}-instance-nested-input`,
    );
  }

  generateOperationNestedInputTypeName(
    entity: Entity,
    operation: Mutation | Subscription,
  ): string {
    const typeName = this.generateEntityTypeName(entity);
    return generateTypeNamePascalCase(
      `${operation.name}-${typeName}-nested-input`,
    );
  }

  generateOperationOutputTypeName(
    entity: Entity,
    operation: Mutation | Subscription,
  ): string {
    const typeName = this.generateEntityTypeName(entity);
    return generateTypeNamePascalCase(`${operation.name}-${typeName}-output`);
  }

  generateOperationTypeName(
    entity: Entity,
    operation: Mutation | Subscription,
  ): string {
    const typeName = this.generateEntityTypeName(entity);
    return generateTypeName(`${operation.name}-${typeName}`);
  }

  generateOperationNestedTypeName(
    entity: Entity,
    operation: Mutation | Subscription,
  ): string {
    const typeName = this.generateEntityTypeName(entity);
    return generateTypeName(`${operation.name}-${typeName}-nested`);
  }

  generateOperationByPrimaryAttributeTypeName(
    entity: Entity,
    operation: Mutation | Subscription,
    attribute,
  ): string {
    const typeName = this.generateEntityTypeName(entity);
    const fieldName = this.generateFieldName(attribute);
    return generateTypeName(`${operation.name}-${typeName}-by-${fieldName}`);
  }
}

export const isProtocolGraphQLConfiguration = (
  obj: any,
): obj is ProtocolGraphQLConfiguration => {
  return obj instanceof ProtocolGraphQLConfiguration;
};
