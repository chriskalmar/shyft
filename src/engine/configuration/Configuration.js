import { passOrThrow, isArray } from '../util';

import { isSchema } from '../schema/Schema';

import { languageIsoCodeRegex, LANGUAGE_ISO_CODE_PATTERN } from '../constants';

import { isProtocolConfiguration } from '../protocol/ProtocolConfiguration';
import { isStorageConfiguration } from '../storage/StorageConfiguration';

import * as _ from 'lodash';

export class Configuration {
  constructor(setup = {}) {
    const {
      languages,
      schema,
      protocolConfiguration,
      storageConfiguration,
    } = setup;

    this.setLanguages(languages || [ 'en' ]);

    if (schema) {
      this.setSchema(schema);
    }

    if (protocolConfiguration) {
      this.setProtocolConfiguration(protocolConfiguration);
    }

    if (storageConfiguration) {
      this.setStorageConfiguration(storageConfiguration);
    }
  }

  setLanguages(languages) {
    passOrThrow(
      isArray(languages, true),
      () => 'Configuration.setLanguages() expects a list of language iso codes',
    );

    languages.map(language => {
      passOrThrow(
        languageIsoCodeRegex.test(language),
        () =>
          `Invalid language iso code '${language}' provided (Regex: /${LANGUAGE_ISO_CODE_PATTERN}/)`,
      );
    });

    passOrThrow(
      languages.length === _.uniq(languages).length,
      () => 'Language codes should be unique in the list',
    );

    this.languages = languages;
  }

  getLanguages() {
    return this.languages;
  }

  setSchema(schema) {
    passOrThrow(isSchema(schema), () => 'Configuration expects a valid schema');

    this.schema = schema;
  }

  getSchema() {
    passOrThrow(this.schema, () => 'Configuration is missing a valid schema');

    return this.schema;
  }

  setProtocolConfiguration(protocolConfiguration) {
    passOrThrow(
      isProtocolConfiguration(protocolConfiguration),
      () => 'Configuration expects a valid protocolConfiguration',
    );

    this.protocolConfiguration = protocolConfiguration;

    protocolConfiguration.getParentConfiguration = () => this;
  }

  getProtocolConfiguration() {
    passOrThrow(
      this.protocolConfiguration,
      () => 'Configuration is missing a valid protocolConfiguration',
    );

    return this.protocolConfiguration;
  }

  setStorageConfiguration(storageConfiguration) {
    passOrThrow(
      isStorageConfiguration(storageConfiguration),
      () => 'Configuration expects a valid storageConfiguration',
    );

    this.storageConfiguration = storageConfiguration;

    storageConfiguration.getParentConfiguration = () => this;
  }

  getStorageConfiguration() {
    passOrThrow(
      this.storageConfiguration,
      () => 'Configuration is missing a valid storageConfiguration',
    );

    return this.storageConfiguration;
  }
}

export const isConfiguration = obj => {
  return obj instanceof Configuration;
};
