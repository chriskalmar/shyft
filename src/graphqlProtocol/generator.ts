import * as _ from 'lodash';
import { RELAY_TYPE_PROMOTER_FIELD } from './protocolGraphqlConstants';
import { graphRegistry } from './graphRegistry';
import { ProtocolGraphQL } from './ProtocolGraphQL';
import { ProtocolGraphQLConfiguration } from './ProtocolGraphQLConfiguration';

import { shaper } from 'json-shaper';

import {
  GraphQLObjectType,
  GraphQLSchema,
  GraphQLNonNull,
  GraphQLID,
} from 'graphql';

import { nodeDefinitions, fromGlobalId, toGlobalId } from 'graphql-relay';

import { registerConnection, generateReverseConnections } from './connection';

import { generateListQueries, generateInstanceQueries } from './query';

import { generateMutations } from './mutation';

import { generateActions } from './action';

import { generateSubscriptions } from './subscription';

import { resolveByFindOne } from './resolver';
import { isConfiguration } from '../engine/configuration/Configuration';
import { isEntity } from '../engine/entity/Entity';
import { DataTypeI18n } from '../engine/datatype/dataTypes';
import {
  ACTION_TYPE_MUTATION,
  ACTION_TYPE_QUERY,
} from '../engine/action/Action';
import { isViewEntity } from '../engine/entity/ViewEntity';
import { isShadowEntity } from '../engine/entity/ShadowEntity';
import { generateInstanceUniquenessInputs } from './operation';

export const getTypeForEntityFromGraphRegistry = (entity) => {
  const typeName = entity.graphql.typeName;
  return graphRegistry.types[typeName];
};

// prepare models for graphql
export const extendModelsForGql = (entities) => {
  const protocolConfiguration = ProtocolGraphQL.getProtocolConfiguration() as ProtocolGraphQLConfiguration;

  _.forEach(entities, (entity) => {
    entity.graphql = entity.graphql || {};
    // generate type names for various cases
    entity.graphql.typeName = protocolConfiguration.generateEntityTypeName(
      entity,
    );
    entity.graphql.typeNamePlural = protocolConfiguration.generateEntityTypeNamePlural(
      entity,
    );
    entity.graphql.typeNamePascalCase = protocolConfiguration.generateEntityTypeNamePascalCase(
      entity,
    );
    entity.graphql.typeNamePluralPascalCase = protocolConfiguration.generateEntityTypeNamePluralPascalCase(
      entity,
    );

    const dataShaperMap = {};

    _.forEach(entity.getAttributes(), (attribute) => {
      attribute.gqlFieldName = attribute.primary
        ? 'id'
        : protocolConfiguration.generateFieldName(attribute);

      dataShaperMap[attribute.gqlFieldName] = attribute.name;

      if (attribute.i18n) {
        attribute.gqlFieldNameI18n = protocolConfiguration.generateI18nFieldName(
          attribute,
        );
        dataShaperMap[attribute.gqlFieldNameI18n] = `${attribute.name}.i18n`;

        attribute.gqlFieldNameI18nJson = protocolConfiguration.generateI18nJsonFieldName(
          attribute,
        );
        dataShaperMap[
          attribute.gqlFieldNameI18nJson
        ] = `${attribute.name}.i18n`;
      }
    });

    // forward relay type promoter field as well
    dataShaperMap[RELAY_TYPE_PROMOTER_FIELD] = RELAY_TYPE_PROMOTER_FIELD;

    // generate json shaper - translate schema attribute names to graphql attribute names
    const dataShaper = shaper(dataShaperMap);
    entity.graphql.dataShaper = (data) => {
      return data ? dataShaper(data) : data;
    };
    entity.graphql.dataSetShaper = (set) => {
      return set.map(entity.graphql.dataShaper);
    };

    // remove i18n JSON output mapping so it doesn't overwrite values in mutation inputs
    _.forEach(entity.getAttributes(), (attribute) => {
      if (attribute.i18n) {
        delete dataShaperMap[attribute.gqlFieldNameI18nJson];
      }
    });

    const reverseDataShaper = shaper(_.invert(dataShaperMap));
    entity.graphql.reverseDataShaper = (data) => {
      return data ? reverseDataShaper(data) : data;
    };
  });
};

// get node definitions for relay
const getNodeDefinitions = () => {
  const idFetcher = (globalId, context) => {
    const { type, id } = fromGlobalId(globalId);

    // resolve based on type and id
    const entity = graphRegistry.types[type]
      ? graphRegistry.types[type].entity
      : null;

    if (entity) {
      const resolver = resolveByFindOne(entity, () => id);
      return resolver(null, null, context);
    }

    return null;
  };

  const typeResolver = (obj) => {
    const typeName = obj[RELAY_TYPE_PROMOTER_FIELD];

    return graphRegistry.types[typeName]
      ? graphRegistry.types[typeName].type
      : null;
  };

  return {
    ...nodeDefinitions(idFetcher, typeResolver),
    idFetcher,
  };
};

const registerAction = (action) => {
  const actionName = action.name;
  graphRegistry.actions[actionName] = {
    action,
  };
};

export const registerActions = (actions) => {
  _.forEach(actions, (action) => registerAction(action));
};

export const generateGraphQLSchema = (configuration) => {
  if (!isConfiguration(configuration)) {
    throw new Error(
      'Invalid configuration object provided to generateGraphQLSchema()',
    );
  }

  const schema = configuration.getSchema();
  const protocolConfiguration = configuration.getProtocolConfiguration();

  ProtocolGraphQL.setProtocolConfiguration(protocolConfiguration);

  const { nodeInterface, nodeField, idFetcher } = getNodeDefinitions();

  registerActions(schema.getActions());

  // prepare models for graphql
  extendModelsForGql(schema.getEntities());

  _.forEach(schema.getEntities(), (entity) => {
    const typeName = entity.graphql.typeName;

    const objectType = new GraphQLObjectType({
      name: protocolConfiguration.generateEntityTypeNamePascalCase(entity),
      description: entity.description,
      interfaces: [nodeInterface],

      fields: () => {
        const fields = {
          id: {
            type: new GraphQLNonNull(GraphQLID),
          },
          nodeId: {
            description: 'The node ID of an object',
            type: new GraphQLNonNull(GraphQLID),
            resolve: (obj) => toGlobalId(typeName, obj.id),
          },
        };

        const fieldsReference = {};
        const fieldsI18n = {};

        _.forEach(entity.getAttributes(), (attribute) => {
          if (attribute.hidden || attribute.mutationInput) {
            return;
          }

          // const field = {
          //   description: attribute.description,
          // };

          let attributeType = attribute.type;

          // it's a reference
          if (isEntity(attributeType) || isShadowEntity(attributeType)) {
            const targetEntity = attributeType;
            const primaryAttribute = targetEntity.getPrimaryAttribute();
            attributeType = primaryAttribute.type;

            // const reference = {
            //   description: attribute.description,
            // };

            const targetTypeName = targetEntity.graphql.typeName;

            // reference.type = graphRegistry.types[targetTypeName].type;
            // reference.resolve = resolveByFindOne(
            //   targetEntity,
            //   ({ source }) => source[attribute.gqlFieldName],
            // );

            const reference = {
              description: attribute.description,
              type: graphRegistry.types[targetTypeName].type,
              resolve: resolveByFindOne(
                targetEntity,
                ({ source }) => source[attribute.gqlFieldName],
              ),
            };

            const referenceFieldName = protocolConfiguration.generateReferenceFieldName(
              targetEntity,
              attribute,
            );
            fieldsReference[referenceFieldName] = reference;
          }

          const fieldType = ProtocolGraphQL.convertToProtocolDataType(
            attributeType,
            entity.name,
            false,
          );

          // make it non-nullable if it's required
          // if (attribute.required) {
          //   field.type = new GraphQLNonNull(fieldType);
          // } else {
          //   field.type = fieldType;
          // }

          // use computed value's function as the field resolver
          // if (attribute.resolve) {
          //   field.resolve = attribute.resolve;
          // }

          const field = {
            description: attribute.description,
            type: attribute.required
              ? new GraphQLNonNull(fieldType)
              : fieldType,
            resolve: attribute.resolve ? attribute.resolve : undefined,
          };

          fields[attribute.gqlFieldName] = field;

          if (attribute.i18n) {
            // JSON i18n output
            const i18nJsonFieldType = ProtocolGraphQL.convertToProtocolDataType(
              DataTypeI18n,
              entity.name,
              false,
            );

            const i18nJsonField = {
              type: attribute.required
                ? new GraphQLNonNull(i18nJsonFieldType)
                : i18nJsonFieldType,
              description: `Translations of **\`${attribute.gqlFieldName}\`** in JSON format`,
            };

            fieldsI18n[attribute.gqlFieldNameI18nJson] = i18nJsonField;

            // Object Type i18n output
            const i18nFieldType = new GraphQLObjectType({
              name: protocolConfiguration.generateI18nFieldTypeName(
                entity,
                attribute,
              ),
              description: `Translations of **\`${attribute.gqlFieldName}\`**`,

              fields: () => {
                const languages = configuration.getLanguages();
                const i18nFields = {};

                languages.map((language, langIdx) => {
                  const type =
                    langIdx === 0 && attribute.required
                      ? new GraphQLNonNull(fieldType)
                      : fieldType;

                  i18nFields[language] = {
                    type,
                  };
                });

                return i18nFields;
              },
            });

            fieldsI18n[attribute.gqlFieldNameI18n] = {
              type: attribute.required
                ? new GraphQLNonNull(i18nFieldType)
                : i18nFieldType,
            };
          }
        });

        Object.assign(
          fields,
          fieldsI18n,
          fieldsReference,
          generateReverseConnections(configuration, graphRegistry, entity),
        );

        return fields;
      },
    });

    graphRegistry.types[typeName] = {
      entity,
      type: objectType,
    };

    if (isEntity(entity) || isViewEntity(entity)) {
      registerConnection(graphRegistry, entity);
    }
  });

  generateInstanceUniquenessInputs(graphRegistry);

  // build the query type
  const queryType = new GraphQLObjectType({
    name: 'Query',
    // root: 'The root query type',

    fields: () => {
      const listQueries = generateListQueries(graphRegistry);
      const instanceQueries = generateInstanceQueries(graphRegistry, idFetcher);
      const actions = generateActions(graphRegistry, ACTION_TYPE_QUERY);

      // override args.id of relay to args.nodeId
      nodeField.args.nodeId = nodeField.args.id;
      nodeField.resolve = (_obj, { nodeId }, context, info) =>
        idFetcher(nodeId, context, info);
      delete nodeField.args.id;

      return {
        node: nodeField,
        ...instanceQueries,
        ...listQueries,
        ...actions,
      };
    },
  });

  const mutationType = new GraphQLObjectType({
    name: 'Mutation',
    // root: 'The root mutation type',

    fields: () => {
      const mutations = generateMutations(graphRegistry);
      const actions = generateActions(graphRegistry, ACTION_TYPE_MUTATION);

      return {
        ...mutations,
        ...actions,
      };
    },
  });

  const subscriptionType = new GraphQLObjectType({
    name: 'Subscription',
    // root: 'The root subscription type',

    fields: () => {
      const subscriptions = generateSubscriptions(graphRegistry);

      return subscriptions;
    },
  });

  // put it all together into a graphQL schema
  return new GraphQLSchema({
    query: queryType,
    mutation: mutationType,
    subscription: subscriptionType,
  });
};
