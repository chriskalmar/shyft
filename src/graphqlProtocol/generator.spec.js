import { generateGraphQLSchema } from './generator';
import { ProtocolGraphQLConfiguration } from './ProtocolGraphQLConfiguration';
import { Configuration } from '../engine/configuration/Configuration';
import {
  DataTypeBoolean,
  DataTypeID,
  DataTypeInteger,
  DataTypeString,
} from '../engine/datatype/dataTypes';
import { Entity } from '../engine/entity/Entity';
import { Schema } from '../engine/schema/Schema';
import { StorageType } from '../engine/storage/StorageType';
import { StorageDataType } from '../engine/storage/StorageDataType';
// import { StorageTypeMemory } from '../memory-connector/StorageTypeMemory';

// const {
//   where: permissionWhere,
//   lookupPermissionEntity,
// } = await buildActionPermissionFilter(
//   permission,
//   userId,
//   userRoles,
//   action,
//   input,
//   context
// );

export const generateTestSchema = async entities => {
  const testEntity = new Entity({
    name: 'FilteredEntityName',
    description: 'Just some description',
    attributes: {
      id: {
        // type: DataTypeInteger,
        type: DataTypeID,
        description: 'ID of user',
        primary: true,
      },
      firstName: {
        type: DataTypeString,
        description: 'First name',
      },
      lastName: {
        type: DataTypeString,
        description: 'Last name',
      },
      isActive: {
        type: DataTypeBoolean,
        description: 'User has been active within this month',
      },
    },
  });

  const StorageDataTypeAny = new StorageDataType({
    name: 'StorageDataTypeAny',
    description: 'Just some description',
    nativeDataType: 'text',
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    serialize() {},
    capabilities: ['lt', 'lte', 'gt', 'gte'],
  });

  const StorageDataTypeText = new StorageDataType({
    name: 'StorageDataTypeText',
    description: 'Just some description',
    nativeDataType: 'text',
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    serialize() {},
    capabilities: ['in', 'lt', 'lte', 'gt', 'gte', 'starts_with', 'ends_with'],
  });

  // const StorageDataTypeBoolean = new StorageDataType({
  //   name: 'StorageDataTypeBoolean',
  //   description: 'Data type representing a boolean value',
  //   nativeDataType: Boolean,
  //   isSortable: true,
  //   serialize: value => !!value,
  // });

  // const StorageDataTypeNumber = new StorageDataType({
  //   name: 'StorageDataTypeNumber',
  //   description: 'Data type representing a numeric value',
  //   nativeDataType: Number,
  //   isSortable: true,
  //   serialize: Number,
  // });

  const StorageDataTypeString = new StorageDataType({
    name: 'StorageDataTypeString',
    description: 'Data type representing a text value',
    nativeDataType: String,
    isSortable: true,
    serialize: String,
  });

  /* eslint-disable no-console */
  const TestStorage = new StorageType({
    name: 'TestStorage',
    description: 'Just some description',
    findOne(entity, id, args, context) {
      console.log('Storage findOne : ', { entity, id, args, context });
      return null;
    },
    findOneByValues(entity, arg, context) {
      console.log('Storage findOneByValues :', {
        entity,
        arg,
        context,
      });
      return null;
    },
    find(entity, args, context, parentConnection) {
      console.log('Storage find : ', {
        entity,
        args,
        context,
        parentConnection,
      });
      // exp;
      return null;
    },
    count(entity, args, context, parentConnection) {
      console.log('Storage count : ', {
        entity,
        args,
        context,
        parentConnection,
      });
    },
    mutate(entity, id, input, entityMutation, context) {
      console.log('Storage mutate : ', {
        entity,
        id,
        input,
        entityMutation,
        context,
      });
      return null;
    },
    checkLookupPermission(entity, where, context) {
      console.log('Storage checkLookupPermission : ', {
        entity,
        where,
        context,
      });
      return true;
    },
  });
  /* eslint-enable no-console */

  TestStorage.addDataTypeMap(DataTypeID, StorageDataTypeString);
  TestStorage.addDataTypeMap(DataTypeInteger, StorageDataTypeAny);
  TestStorage.addDataTypeMap(DataTypeBoolean, StorageDataTypeAny);
  TestStorage.addDataTypeMap(DataTypeString, StorageDataTypeText);

  const schema = new Schema({
    defaultStorageType: TestStorage,
    // defaultStorageType: StorageTypeMemory,
    // defaultActionPermissions: devOnlyPermission,
    // permissionsMap,
    entities: entities ? [...entities, testEntity] : [testEntity],
  });

  const configuration = new Configuration({
    languages: ['en'],
    schema,
    protocolConfiguration: new ProtocolGraphQLConfiguration(),
  });

  // console.log('Schema', { schema });
  return { configuration, schema };
};

let configuration;

beforeAll(async () => {
  const setup = await generateTestSchema();
  configuration = setup.configuration;
});

describe('generator', () => {
  describe('test', () => {
    it('should render GraphQL Schema', () => {
      const graphqlSchema = generateGraphQLSchema(configuration);
      expect(typeof graphqlSchema).toEqual('object');
      // console.log('graphqlSchema', { graphqlSchema });
      expect(graphqlSchema).toMatchSnapshot();
    });
  });
});
