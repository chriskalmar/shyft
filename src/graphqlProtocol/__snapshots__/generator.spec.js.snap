// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generator test should render GraphQL Schema 1`] = `
GraphQLSchema {
  "__allowedLegacyNames": Array [],
  "__validationErrors": undefined,
  "_directives": Array [
    "@include",
    "@skip",
    "@deprecated",
  ],
  "_implementations": Object {
    "Node": Array [
      "FilteredEntityName",
    ],
  },
  "_mutationType": "Mutation",
  "_possibleTypeMap": Object {},
  "_queryType": "Query",
  "_subscriptionType": undefined,
  "_typeMap": Object {
    "Boolean": "Boolean",
    "CreateFilteredEntityNameInput": "CreateFilteredEntityNameInput",
    "CreateFilteredEntityNameInstanceInput": "CreateFilteredEntityNameInstanceInput",
    "CreateFilteredEntityNameInstanceNestedInput": "CreateFilteredEntityNameInstanceNestedInput",
    "CreateFilteredEntityNameNestedInput": "CreateFilteredEntityNameNestedInput",
    "CreateFilteredEntityNameOutput": "CreateFilteredEntityNameOutput",
    "Cursor": "Cursor",
    "DeleteFilteredEntityNameByIdInput": "DeleteFilteredEntityNameByIdInput",
    "DeleteFilteredEntityNameInput": "DeleteFilteredEntityNameInput",
    "DeleteFilteredEntityNameOutput": "DeleteFilteredEntityNameOutput",
    "FilteredEntityName": "FilteredEntityName",
    "FilteredEntityNameConnection": "FilteredEntityNameConnection",
    "FilteredEntityNameEdge": "FilteredEntityNameEdge",
    "FilteredEntityNameFilter": "FilteredEntityNameFilter",
    "ID": "ID",
    "Int": "Int",
    "Mutation": "Mutation",
    "Node": "Node",
    "PageInfo": "PageInfo",
    "Query": "Query",
    "String": "String",
    "UpdateFilteredEntityNameByIdInput": "UpdateFilteredEntityNameByIdInput",
    "UpdateFilteredEntityNameInput": "UpdateFilteredEntityNameInput",
    "UpdateFilteredEntityNameInstanceInput": "UpdateFilteredEntityNameInstanceInput",
    "UpdateFilteredEntityNameInstanceNestedInput": "UpdateFilteredEntityNameInstanceNestedInput",
    "UpdateFilteredEntityNameNestedInput": "UpdateFilteredEntityNameNestedInput",
    "UpdateFilteredEntityNameOutput": "UpdateFilteredEntityNameOutput",
    "__Directive": "__Directive",
    "__DirectiveLocation": "__DirectiveLocation",
    "__EnumValue": "__EnumValue",
    "__Field": "__Field",
    "__InputValue": "__InputValue",
    "__Schema": "__Schema",
    "__Type": "__Type",
    "__TypeKind": "__TypeKind",
  },
  "astNode": undefined,
  "extensionASTNodes": undefined,
  "extensions": undefined,
}
`;
